File: src/main/java/org/dreamexposure/ticketbird/module/status/StatusChanger.java
package org.dreamexposure.ticketbird.module.status;

import discord4j.core.GatewayDiscordClient;
import discord4j.core.object.presence.Activity;
import discord4j.core.object.presence.Presence;
import org.dreamexposure.ticketbird.TicketBird;

import java.util.ArrayList;
import java.util.TimerTask;

import static org.dreamexposure.ticketbird.GitProperty.TICKETBIRD_VERSION;

public class StatusChanger extends TimerTask {
    private final ArrayList<String> statuses = new ArrayList<>();
    private int index;

    private final GatewayDiscordClient client;

    /**
     * Creates the StatusChanger and its Statuses list.
     */
    public StatusChanger(GatewayDiscordClient client) {
        this.client = client;

        statuses.add("TicketBird Bot!");
        statuses.add("=help for help");
        statuses.add("=ticketbird for info");
        statuses.add("Powered by DreamExposure");
        statuses.add("Used on %guCount% guilds!");
        statuses.add("%shards% shards!");
        statuses.add("Version " + TICKETBIRD_VERSION.getValue());
        statuses.add("TicketBird is on Patreon!");
        statuses.add("Share TicketBird!!");
        index = 0;
    }

    @Override
    public void run() {
        String status = statuses.get(index);

        if (status.contains("%guCount%"))
            status = status.replace("%guCount%", client.getGuilds().count().block() + "");
        status = status.replace("%shards%", TicketBird.getShardCount() + "");

        client.updatePresence(Presence.online(Activity.playing(status))).subscribe();

        //Set new index.
        if (index + 1 >= statuses.size())
            index = 0;
        else
            index++;
    }
}

File: src/main/kotlin/org/dreamexposure/ticketbird/service/StatusChanger.kt
package org.dreamexposure.ticketbird.service

import discord4j.core.GatewayDiscordClient
import discord4j.core.`object`.presence.ClientActivity
import discord4j.core.`object`.presence.ClientPresence
import kotlinx.coroutines.reactor.awaitSingle
import kotlinx.coroutines.reactor.awaitSingleOrNull
import kotlinx.coroutines.reactor.mono
import org.dreamexposure.ticketbird.GitProperty
import org.dreamexposure.ticketbird.logger.LOGGER
import org.dreamexposure.ticketbird.utils.GlobalVars.DEFAULT
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Duration
import java.util.concurrent.atomic.AtomicInteger

@Component
class StatusChanger(private val client: GatewayDiscordClient) : ApplicationRunner {
    private val index = AtomicInteger(0)

    private val statuses = listOf(
        "/ticketbird for info & help",
        "Do you know what a square is?",
        "Version {version}",
        "Now supports Slash Commands!",
        "Trans rights are human rights",
        "Sorting tickets in {guild_count} guilds!",
        "Proudly written in Kotlin using Discord4J",
        "Slava Ukraini!",
    )

    private fun update() = mono {
        val currentIndex = index.get()
        //Update index
        if (currentIndex + 1 >= statuses.size)
            index.lazySet(0)
        else
            index.lazySet(currentIndex + 1)

        //Get status we want to change to
        var status = statuses[currentIndex]
            .replace("{version}", GitProperty.TICKETBIRD_VERSION.value)

        if (status.contains("{guild_count}")) {
            val count = client.guilds.count().awaitSingle()
            status = status.replace("{guild_count}", "$count")
        }


        client.updatePresence(ClientPresence.online(ClientActivity.playing(status))).awaitSingleOrNull()
    }

    override fun run(args: ApplicationArguments?) {
        Flux.interval(Duration.ofMinutes(5))
            .flatMap { update() }
            .doOnError { LOGGER.error(DEFAULT, "StatusChanger | Processor failure ", it) }
            .onErrorResume { Mono.empty() }
            .subscribe()
    }
}
